### PR TITLE
protocol/tx: add extensible tx hashing

### PR DESCRIPTION
**Pull request by bobg**
This new package implements a change to transaction hashing that is more flexible and futureproof. Transactions are converted to new data structures that are intended eventually to replace the types in `protocol/bc`. This transitional code is intended to protect against hash values changing when we reach that point.

The spec for these changes is in `docs/protocol/specifications/txgraph-draft.md`.

Still to come (in follow-on PRs): the code that integrates this new package with the existing code in `protocol/bc`; and integration of `txgraph-draft.md` into the spec proper.

### CODE REVIEW

**Comment by bobg** [comment]
PTAL

**Comment by bobg** [comment]
Anyone reviewing this code and wishing to see how it will integrate, please refer to the `txgraph-integ` branch for now.

**Comment by jbowens** [comment]
👀

**Comment by tessr** [comment]
I feel like I should buy Jackson a boba for reviewing this

**Comment by bobg** [comment]
> I feel like I should buy Jackson a boba for reviewing this

Or review [_his_ giant PR](https://github.com/chain/chain/pull/428).

**Comment by jbowens** [comment]
> Or review his giant PR.

I think @erykwalder is on it :). Reviewing this is no worry at all

File: protocol/tx/transaction.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
+package tx
+
+import (
+	"chain/crypto/sha3pool"
+	"chain/protocol/bc"
+	"fmt"
```
Comment (jbowens) [inline protocol/tx/transaction.go]: nit: import grouping

File: protocol/tx/entry.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	return hash, nil
+}
+
+func writeForHash(w io.Writer, c interface{}) error {
+	switch v := c.(type) {
+	case byte:
+		_, err := w.Write([]byte{v})
+		return err
+
```
Comment (jbowens) [inline protocol/tx/entry.go]: nit: whitespace dividing cases here, but not elsewhere

**Comment by jbowens** [comment]
lgtm. haven't checked for complete equality between the spec and implementation but don't think that should necessarily block landing this

File: protocol/tx/data.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// the entry itself stores a 32-byte hash of the underlying data,
+// which may be of any length.
+// It is the responsibility of the transport layer
+// to provide the underlying data
+// alongside the actual transaction, if necessary.
+// The data need not be made available to all parties;
+// it is fine to keep it confidential.
+//
+// Note that the body of this entry is a hash (of the underlying data);
+// the body_hash is a hash of that hash.
```
Comment (tessr) [inline protocol/tx/data.go]: I think it means that the hashed representation is a hash of `body` (which is a hash of the underlying data). I'll try to improve this comment.

File: protocol/tx/map.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	var firstSpendID *entryRef
+	muxSources := make([]valueSource, len(tx.Inputs))
+
+	maybeAddInputRefdata := func(inp *bc.TxInput) (ref entryRef, err error) {
+		if len(inp.ReferenceData) == 0 {
+			return
+		}
+		ref, _, err = addEntry(newData(hashData(inp.ReferenceData)))
+		return
+	}
```
Comment (tessr) [inline protocol/tx/map.go]: Agreed.

**Pull request merged**